spec/abci++/v4.md

```diff
@@ -0,0 +1,200 @@
+# Tendermint v3 Markdown pseudocode
+
+This is a single-threaded implementation of ABCI++,
+with an optimization for the ProcessProposal phase.
+Namely, processing of the header and the block data is separated into two different functions.
+
+### Initialization
+
+```go
+h_p ← 0
+round_p ← 0
+step_p is one of {propose, prevote, precommit}
+decision_p ← Vector()
+lockedValue_p ← nil
+validValue_p ← nil
+validRound_p ← -1
+```
+
+### StartRound(round)
+
+```go
+function startRound(round) {
+    round_p ← round
+    step_p ← propose
+    if proposer(h_p, round_p) = p {
+        if validValue_p != nil {
+            proposal ← validValue_p
+        } else {
+            txdata ← mempool.GetBlock()
+            // getUnpreparedBlockProposal fills in header
+            unpreparedProposal ← getUnpreparedBlockProposal(txdata)
+            proposal ← ABCI.PrepareProposal(unpreparedProposal)
+        }
+        broadcast ⟨PROPOSAL, h_p, round_p, proposal, validRound_p⟩
+    } else {
+        schedule OnTimeoutPropose(h_p,round_p) to be executed after timeoutPropose(round_p)
+    }
+}
+```
+
+### ReceiveProposal
+
+In the case where the local node is not locked on any round, the following is ran:
+
+```go
+upon ⟨PROPOSAL, h_p, round_p, v, −1) from proposer(h_p, round_p) while step_p = propose do {
+    if valid(v) ∧ ABCI.VerifyHeader(h_p, v.header) ∧ (lockedRound_p = −1 ∨ lockedValue_p = v) {
+        // We fork process proposal into a parallel process
+        Fork ABCI.ProcessProposal(h_p, v)
```

Comment:
We discussed in slack but just noting here the concern with this is that we could end up with a POLKA for a block that turns out to be invalid (ie. fails ProcessProposal) and this could have liveness consequences. 

So we need to carefully work through the consequences of this and adjust as necessary or else just suffer the latency cost of completing ProcessProposal before prevoting.

Reply:
It might be fine prevoting for v without waiting ProcessProposal(v) to return, but we should probably not set locked/valid value if ProcessProposal(v) hasn't returned successfully.  Therefore, we ensure that validValue/lockedValue is never set to invalid block, and this ensures that invalid block is never committed.